pointercrate-user/src/auth/get.rs

```diff
@@ -46,9 +60,117 @@ impl AuthenticatedUser {
         Ok(user)
     }
 
+    pub async fn oauth2_callback(code: &str, existing_id: Option<i32>, connection: &mut PgConnection) -> Result<AuthenticatedUser> {
+        info!("We are expected to perform Google OAuth2 authentication");
+
+        let client = reqwest::Client::new();
+
+        let response = client
+            .post("https://oauth2.googleapis.com/token")
+            .form(&[
+                ("code", code),
+                ("client_id", &config::google_client_id()),
+                ("client_secret", &config::google_client_secret()),
+                ("redirect_uri", &config::google_redirect_uri()),
+                ("grant_type", "authorization_code"),
+            ])
+            .send()
+            .await
+            .map_err(|_| CoreError::Unauthorized)?;
+
+        let response: GoogleTokenResponse = response.json().await.map_err(|_| CoreError::Unauthorized)?;
+
+        // We can safely disable all validation here, as Google recommends to not
+        // validate a fresh token, as it is guaranteed to be valid.
+        //
+        // https://developers.google.com/identity/openid-connect/openid-connect#obtainuserinfo
+        let key = DecodingKey::from_secret(&[]);
+        let mut validation = Validation::new(Algorithm::HS256);
+        validation.insecure_disable_signature_validation();
+        validation.validate_exp = false;
+        validation.validate_nbf = false;
+        validation.validate_aud = false;
+
+        let user_info = jsonwebtoken::decode::<GoogleUserInfo>(&response.id_token, &key, &validation)
+            .map_err(|_| CoreError::Unauthorized)?
+            .claims;
+
+        match User::by_google_account(&user_info.id, connection).await {
+            Ok(user) => Ok(AuthenticatedUser {
+                user,
+                password_hash: None,
+                google_account_id: Some(user_info.id),
+                email_address: Some(user_info.email),
+            }),
+            Err(UserError::UserNotFoundGoogleAccount { .. }) => {
+                if let Some(id) = existing_id {
+                    let user = AuthenticatedUser::by_id(id, connection).await?;
+
+                    if user.google_account_id.is_some() {
+                        return Err(CoreError::Unauthorized.into());
+                    }
+
+                    let updated_user = sqlx::query!(
+                        "UPDATE members SET google_account_id = $1, email_address = ($2::text)::email WHERE member_id = $3 RETURNING member_id",
```

Comment:
I think here is the problem why this flow gives you an invalid access token: this query is missing a ` SET password_hash = NULL`. Right now in line 126 onward, you're constructing an `AuthenticatedUser` with `password_hash: None`. This means that if you call `generate_access_token()` on the user returned here, it will only sign the token with the server secret. However, the database will still contain the user's old password hash. This means that when it comes to validating the token, it will retrieve an `AuthenticatedUser` with `password_hash = hash of password from before account was linked`, and thus it tries to validate the token using a secret that's a combination of server secret and password hash. But that's the wrong secret, because we signed the token with only the server secret!